File: bin/collapse_concatenated_reads.R
library(logger)
log_threshold(WARN)
log_appender(appender_console)

load_library <- function(library_name) {
  output <- capture.output({
    suppressWarnings({
      library(library_name, character.only = TRUE)
    })
  }, type = "message")
  
  # Separate warnings from messages
  warnings <- warnings()
  
  # Log messages
  if(length(output) > 0) {
    log_info(paste("Message from", library_name, ":", paste(output, collapse = "; ")))
  }
  
  # Log warnings
  if(length(warnings) > 0) {
    log_warn(paste("Warning from", library_name, ":", paste(warnings, collapse = "; ")))
  }
}

load_library("argparse")
load_library("dplyr")
load_library("magrittr")
load_library("stringr")

parser <- ArgumentParser(description='Handles denoising of sequences formed by joining reads through concatenation by DADA2')
parser$add_argument('--clusters', type="character", required=TRUE, help="RDS Clusters from DADA2. This is the main output from the DADA module.")
parser$add_argument('--log-level', type="character", default = "INFO", help = "Log level. Default is INFO.")

args <- parser$parse_args()

# Set the logging threshold 
log_level_arg <- match.arg(args$log_level, c("DEBUG", "INFO", "WARN", "ERROR", "FATAL"))
log_threshold(log_level_arg)

# Log the arguments parsed
args_string <- paste(sapply(names(args), function(name) {
  paste(name, ":", args[[name]])
}), collapse = ", ")

log_debug(paste("Arguments parsed successfully:", args_string))

## FOR DEBUGGING
# args=list()
# setwd("/home/bpalmer/Documents/GitHub/mad4hatter/work/e8/fbf5c2d9377a2e58bb236de5c872d7")
# args$clusters="dada2.clusters.txt"

# load the output from the dada2 process (allele_data saved into dada2.clusters.RDS)
clusters=NULL
if (grepl(".RDS|.rds", args$clusters)) {
  clusters=readRDS(args$clusters)
} else {
  clusters=read.table(args$clusters, header=T)
}

## I. Check for non overlapping sequences.

# make data frame clusters that has the distinct asvs and records in which row (sample/locus/allele in clusters) it showed
clusters %<>% dplyr::mutate(seqid=row_number())

clusters.1 = clusters %>%
  group_by(locus,asv) %>%
  dplyr::mutate(seqid=paste(seqid,collapse=";")) %>%
  select(locus,asv,seqid) %>%
  distinct()

# make a list of sequences
sequences = clusters.1$asv
# identify the sequences that didn't overlap as those that have 10 N's, which is how they are concatenated in merge in dada2
non_overlaps_idx <- which(str_detect(sequences, paste(rep("N", 10), collapse = "")))


# if there are any sequences that do not overlap,
# see if they can be collapsed and sum their
# counts

# here we will collapse all the non overlapping sequences that only differed in the length of the reads on either side
# the thinking is that those differences are driven by differences in quality of reads rather than actual variation


# Note for Brian: in df_aln we would have only unique sequences but there's no collapsing of the truncated sequences so those are repeated
if (length(non_overlaps_idx) > 0) {

  # extract the non overlapping sequences
  non_overlaps <- sequences[non_overlaps_idx]
  # split the two ends of the sequence
  non_overlaps_split = strsplit(non_overlaps, paste(rep("N", 10), collapse = ""))

  # put both ends of the sequence and the positions in the sequences list in a data frame
  df_non_overlap <- data.frame(
    column_indexes = non_overlaps_idx,  # note for Brian: why are these called column_indexes if they are from rows?
    left_sequences = sapply(non_overlaps_split, "[[", 1),
    right_sequences = sapply(non_overlaps_split, "[[", 2)
  )

  # extract and order the left sequences and order them alphabetically (and keep the indexes in sequence as names)
  ordered_left_sequences <- df_non_overlap %>%
    arrange(left_sequences, str_length(left_sequences)) %>%
    pull(left_sequences, name = column_indexes)

  idx <- 1
  # store modified left sequences in this list
  processed_left_sequences <- c()

  # for each sequence find any others that start with the same sequence
  while (idx <= length(ordered_left_sequences)) {
    base_sequence <- ordered_left_sequences[idx]
    # 'x' is all sequences that match the base sequence. they should
    # already be together because of the prerequiste seqtab sorting above
    # sequences were sorted alphabetically so matching ones will be together, and then sorted by length so the shortest will come first
    # note that some left sequences will be identical because the differences in the asv are in the right sequence
    x <- ordered_left_sequences[
      startsWith(ordered_left_sequences, base_sequence)]

    if (length(x) > 1) {     # Note for Brian: why only if you have 2 and if the difference is 1? I'm temporarily changing to more than 1 and at most 3, but that's an arbitrary number to not get rid of true trs
    # if there are sequences that are longer than the first by more than 3 bases then those are not analyzed/modified
    #if (length(t) == 2 && (max(t) - min(t) == 1)) {
      t <- nchar(x)
      idx.close = which((t-t[1])<4)
      x <- substr(x[idx.close], 1, t[1])
    }

    processed_left_sequences <- c(processed_left_sequences, x)

    # update index
    idx <- idx + length(x)
  }

  # Now do the same for the right hand sequences!

  # reverse the sequences for easier comparison
  tmp <- data.frame(
    column_indexes = df_non_overlap$column_indexes,
    reversed = sapply(lapply(strsplit(df_non_overlap$right_sequences, NULL), rev), paste, collapse="")
  )
  # extract and order the reversed right sequences and order them alphabetically (and keep the indexes in sequence as names)
  ordered_reversed_right_sequences <- tmp %>%
    arrange(reversed, str_length(reversed)) %>%
    pull(reversed, name = column_indexes)
  # start idx to loop over sequences
  idx <- 1
  # store modified sequences in this list
  processed_right_sequences <- c()

  while (idx <= length(ordered_reversed_right_sequences)) {
    base_sequence <- ordered_reversed_right_sequences[idx]

    x <- ordered_reversed_right_sequences[
      startsWith(ordered_reversed_right_sequences, base_sequence)]

    # if there are more than 1 sequences , take the ones that differ by less than 4 bases and shorten to the length of the first one, if there are longer ones, leave out for the next round
    if (length(x) >1) {
      t <- nchar(x)
      idx.close = which((t-t[1])<4)
      x <- substr(x[idx.close], 1, t[1])
    }

    #append the sequences, modified if so
    processed_right_sequences <- c(processed_right_sequences, x)

    # update index
    idx <- idx + length(x)
  }

  # reverse the sequences
  processed_right_sequences <- sapply(lapply(strsplit(processed_right_sequences, NULL), rev), paste, collapse="")

  # re-order the sequences by their indexes (ie. [1..n])
  processed_left_sequences <- processed_left_sequences[
    order(as.integer(names(processed_left_sequences)))]

  processed_right_sequences <- processed_right_sequences[
    order(as.integer(names(processed_right_sequences)))]

  # combine processed sequences
  df_non_overlap <- cbind(df_non_overlap,
                          processed_left_sequences = processed_left_sequences,
                          processed_right_sequences = processed_right_sequences
  )

  df_non_overlap %<>%
    mutate(combined = paste0(processed_left_sequences,"N",processed_right_sequences))

  write.table(df_non_overlap,file="concatenated_sequences.txt",quote=F,sep="\t",col.names=T,row.names=F)

  # finally write back to sequences
  sequences[df_non_overlap$column_indexes] <- df_non_overlap$combined
}

clusters.1$asv <- sequences
clusters.1 %<>%
  group_by(locus,asv) %>%
  dplyr::mutate(seqid=strsplit(seqid, ";")) %>%
  tidyr::unnest(cols=c(seqid)) %>%
  dplyr::mutate(seqid=as.integer(seqid))

clusters %<>%
    select(-c("asv")) %>%
    inner_join(clusters.1, by=c("locus","seqid")) %>%
    group_by(sampleID,locus,asv) %>%
    dplyr::mutate(reads=sum(reads)) %>%
    group_by(sampleID,locus) %>%
    mutate(norm.reads.locus = reads/sum(reads))%>%
    mutate(n.alleles = n()) %>%
    ungroup() %>%
	  distinct()

clusters$seqid=NULL
write.table(clusters,file="clusters.concatenated.collapsed.txt",quote=F,sep="\t",col.names=T,row.names=F)

File: bin/collapse_concatenated_reads.py
#!/usr/bin/env python3
"""
Collapse concatenated reads.

Usage:
    python collapse_concatenated_reads.py --clusters path/to/asv_table.tsv
"""

import argparse
import pandas as pd
import sys

# TODO: add unit tests using nftest
# TODO: test full pipeline 
# TODO: remove old R script

def main():
    parser = argparse.ArgumentParser(description='Collapse concatenated reads with 10 Ns')
    parser.add_argument('--clusters', required=True, help='Path to clustered ASV table file')
    parser.add_argument('--output', help='Output file path (default: clusters.concatenated.collapsed.txt)')
    
    args = parser.parse_args()
    
    try:
        # Read the ASV table
        print(f"Reading clusters from: {args.clusters}", file=sys.stderr)
        df = pd.read_csv(args.clusters, sep='\t')
        
        # Check if required columns exist
        required_columns = ['asv', 'sampleID', 'locus', 'reads']
        missing_columns = [col for col in required_columns if col not in df.columns]
        if missing_columns:
            print(f"Error: Missing required columns: {missing_columns}", file=sys.stderr)
            print(f"Available columns: {list(df.columns)}", file=sys.stderr)
            sys.exit(1)
        
        # Filter for ASVs with exactly 10 Ns
        print("Processing ASVs with exactly 10 Ns...", file=sys.stderr)
        concatenated_df = df[df['asv'].str.count('N') == 10]
        unconcatenated_df = df[df['asv'].str.count('N') != 10]
        
        print(f"Original table: {len(df)} rows", file=sys.stderr)
        print(f"Concatenated ASVs: {len(concatenated_df)} rows", file=sys.stderr)
        print(f"Non-concatenated ASVs: {len(unconcatenated_df)} rows", file=sys.stderr)
        
        # Process concatenated ASVs if any exist
        if len(concatenated_df) > 0:
            collapsed_df = collapse_asvs(concatenated_df)
            print(f"Collapsed concatenated ASVs: {len(collapsed_df)} rows", file=sys.stderr)
    
            # Combine with non-concatenated ASVs (leave unconcatenated rows untouched)
            final_df = pd.concat([unconcatenated_df, collapsed_df], ignore_index=True)
            
            # Calculate normalized reads and allele counts without re-aggregating non-concatenated ASVs
            grp = final_df.groupby(['sampleID', 'locus'], sort=False)
            final_df = final_df.copy()
            final_df['norm.reads.locus'] = final_df['reads'] / grp['reads'].transform('sum')
            final_df['n.alleles'] = grp['asv'].transform('count')

        else:
            final_df = unconcatenated_df

        # Output results with all columns ordered correctly
        output_file = args.output if args.output else "clusters.concatenated.collapsed.txt"
        desired_cols = ['sampleID', 'locus', 'asv', 'reads', 'allele', 'norm.reads.locus', 'n.alleles']
        final_df = final_df[desired_cols]
        final_df.to_csv(output_file, sep='\t', index=False)
        print(f"Output saved to: {output_file}", file=sys.stderr)
        print(f"Final table: {len(final_df)} rows", file=sys.stderr)
            
    except FileNotFoundError:
        print(f"Error: File '{args.clusters}' not found", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

def collapse_asvs(concat_clusters):
    # Work on a copy to avoid SettingWithCopyWarning
    concat_clusters = concat_clusters.copy()
    concat_clusters.loc[:, 'left'] = concat_clusters['asv'].str.split('NNNNNNNNNN').str[0]
    concat_clusters.loc[:, 'right'] = concat_clusters['asv'].str.split('NNNNNNNNNN').str[1]
    concat_clusters.loc[:, 'len_left'] = concat_clusters['left'].str.len()
    concat_clusters.loc[:, 'len_right'] = concat_clusters['right'].str.len()
    # Avoid groupby.apply deprecation by iterating groups explicitly
    outputs = []
    for (_, _), g in concat_clusters.groupby(['sampleID', 'locus'], sort=False):
        outputs.append(collapse_group(g))
    if len(outputs) == 0:
        return pd.DataFrame(columns=['sampleID','locus','asv','reads','allele'])
    return pd.concat(outputs, ignore_index=True)
     
def collapse_group(df):
    """Collapse sequences within a group by finding common prefixes and suffixes."""
    options = []
    for _, row in df.iterrows():
        options.append({'left': row.left, 'right': row.right})

    new_lefts = []
    new_rights = []
    
    for _, row in df.iterrows(): 
        new_left = row.left
        new_right = row.right
        
        for option in options: 
            # Check if sequences can be collapsed based on prefix/suffix matching
            left_match = (row.left.startswith(option['left']) or 
                         option['left'].startswith(row.left))
            right_match = (row.right.endswith(option['right']) or 
                          option['right'].endswith(row.right))
            
            if left_match and right_match:
                # Take the shorter sequence for collapsing
                if len(option['left']) < len(new_left):
                    new_left = option['left']
                if len(option['right']) < len(new_right):
                    new_right = option['right']
                    
        new_lefts.append(new_left)
        new_rights.append(new_right)
    
    df['new_left'] = new_lefts
    df['new_right'] = new_rights
    
    # Group by the collapsed sequences and sum reads
    df_new = df.groupby(['sampleID', 'locus', 'new_left', 'new_right']).reads.sum().reset_index()
    df_new['asv'] = df_new.new_left + 'NNNNNNNNNN' + df_new.new_right
    df_new = df_new[['sampleID', 'locus', 'asv', 'reads']]
    
    # Add allele column
    allele_df = df[['locus','asv','allele']].drop_duplicates()
    df_new = df_new.merge(allele_df, on=['locus','asv'], how='left')
    # Convert concat to single N 
    df_new['asv'] = df_new.asv.str.replace('NNNNNNNNNN','N')
    return df_new

if __name__ == "__main__":
    main()